main/exercice/exercise.lib.php

```diff
@@ -436,124 +436,73 @@ function showQuestion(
+                        // display the blank word
+                        $correctItem = $listAnswerInformations["tabwords"][$i];
+                        $correctItemRegexp = $correctItem;
+                        // replace / with \/ to allow the preg_replace bellow and all the regexp char
+                        $correctItemRegexp = FillBlanks::getRegexpProtected($correctItemRegexp);
+                        if (isset($studentAnswerList[$i])) {
+                            // student already start this test and this question
+                            // fill the blank with his previous answers
+                            // may be "" if student did the question, but not fill the blanks
+                            $correctItem = $studentAnswerList[$i];
+                        }
+                        $attributes["style"] = "width:".$listAnswerInformations["tabinputsize"][$i]."px";
+                        $answer .= FillBlanks::getFillTheBlankHtml($separatorStartRegexp, $separatorEndRegexp, $correctItemRegexp, $questionId, $correctItem, $attributes, $answer, $listAnswerInformations, $displayForStudent, $i);
                     }
-                    $answer .= $tab_comments[$i];
+                    // display the last common word
+                    $answer .= $listAnswerInformations["commonwords"][$i];
                 } else {
-                    // display exercise with empty input fields
-                    // every [xxx] are replaced with an empty input field
-                    foreach ($correct_answer_list[0] as $item) {
-                        $size = strlen($item);
-                        $attributes['class'] = detectInputAppropriateClass($size);
-                        $answer = str_replace(
-                            $item,
-                            Display::input('text', "choice[$questionId][]", '', $attributes),
-                            $answer
-                        );
+                    // display empty [input] with the right width for student to fill it
+                    $separatorStartRegexp = FillBlanks::escapeForRegexp($listAnswerInformations['blankseparatorstart']);
+                    $separatorEndRegexp = FillBlanks::escapeForRegexp($listAnswerInformations['blankseparatorend']);
+                    $answer = "";
+                    for ($i = 0; $i < count($listAnswerInformations["commonwords"]) - 1; $i++) {
+                        // display the common words
+                        $answer .= $listAnswerInformations["commonwords"][$i];
+                        // display the blank word
+                        $attributes["style"] = "width:".$listAnswerInformations["tabinputsize"][$i]."px";
+                        $correctItem = $listAnswerInformations["tabwords"][$i];
+                        $correctItemRegexp = $correctItem;
+                        // replace / with \/ to allow the preg_replace bellow and all the regexp char
```

Comment:
I've tested this carefully, and here are the limitation
(i'll put in the doc)
"Pour éviter des soucis d'affichage, les caractères suivants \ et " sont supprimés automatiquement des crochets. Les crochets ne doivent pas être vides.
Les espaces de débuts et de fin de mot entre crochets sont automatiquement supprimés ([   mot   ] => [mot])."